### PR TITLE
Add spec for `brew` package URLs

### LINKED ISSUE #254: Proposal: `brew` type
First of all, thank you for creating and maintaining this spec! I think it's a great idea, and I'm looking forward to using (and contributing to) it. 

I'm opening this to start a rough sketch of a purl type for [Homebrew](https://brew.sh) formulae. 

Some initial bits:

* The default repository should probably be `homebrew/core`
   * Homebrew's tap syntax is `{org}/homebrew-{tap}` for GitHub taps, so `homebrew/core` corresponds to `https://github.com/homebrew/homebrew-core`
* The `name` is the Formula name 
   * Formula names can contain `@` (e.g. `llvm@17`), so this will need to be percent-encoded
   * Formulae with `name@version` are typically "fixed" versions,
* The `version` is the Formula version
* Qualifiers:
   * `revision` (when the formula itself is patched) (is this worth including?)
   * `version_scheme` (when the version ordering is changed, similar to epochs in Python) (is this worth including?)
   * Bottle checksum/tag/rebuild (it may not make sense to include these, since resolving the formula itself resolves them)
   * Other things I'm forgetting?

Things not covered (yet):
* Casks

So, some rough examples:

```
# equivalent: implicit namespace and explicit
pkg:brew/sqlite@3.43.2
pkg:brew/homebrew/core/sqlite@3.43.2

# third-party tap
pkg:brew/stripe/stripe-cli/stripe@1.18.0

# percent-encoding for AT-formulae (implicit and explicit underlying version)
pkg:brew/openssl%401.1
pkg:brew/openssl%401.1@1.1.1w
```

CC @p-linnane @colindean @SMillerDev as interested parties 🙂

### PR DESCRIPTION
**Pull request by woodruffw**
This adds the `brew` purl type.

Closes #254.

### CODE REVIEW

File: PURL-TYPES.rst [hunk 1]
```diff
@@ -87,6 +87,25 @@ bitnami
       pkg:bitnami/wordpress@6.2.0
       pkg:bitnami/wordpress@6.2.0?arch=arm64
 
+brew
+----
+``brew`` for Homebrew-based packages:
+
+- The default repository (tap) is ``homebrew/core``.
```
Comment (colindean) [inline PURL-TYPES.rst:94]: **suggestion (non-blocking):** 
```suggestion
- The default repository, which is called a "Tap" in Homebrew terminology, is ``homebrew/core``.
```

File: PURL-TYPES.rst [hunk 2]
```diff
@@ -87,6 +87,25 @@ bitnami
...
+
+  - The tap syntax is ``https://github.com/{org}/Homebrew-{tap}``, so
+    ``homebrew/core`` corresponds to the URL ``https://github.com/homebrew/homebrew-core``.
```
Comment (colindean) [inline PURL-TYPES.rst:97]: **suggestion (non-blocking):** 
```suggestion
  - The typical tap identifier expands to the URL ``https://github.com/{org}/Homebrew-{tap}``, so the tap identifier
    ``homebrew/core`` corresponds to the tap URL ``https://github.com/homebrew/homebrew-core``.
```

I'm maybe too wordy here, but for a spec, wordy might be good.

File: PURL-TYPES.rst [hunk 3]
```diff
@@ -87,6 +87,25 @@ bitnami
...
+- The ``name`` is the formula name.
```
Comment (colindean) [inline PURL-TYPES.rst:98]: **suggestion (non-blocking):** Add this within the previous list item:
```suggestion
  - Any git URL is a valid tap URL (see ``tap_url``)
- The ``name`` is the formula name.
```

This clarifies from the start, but maybe the `tap_url` bullet belongs in here?

File: PURL-TYPES.rst [hunk 4]
```diff
@@ -87,6 +87,25 @@ bitnami
...
+- The ``version`` is the formula version.
```
Comment (colindean) [inline PURL-TYPES.rst:99]: **question (non-blocking):** When thinking about the examples below, I think you should clarify  how purl spec should handle version-in-formula formulae.

As in, if I want PostgreSQL 12.17 (current release as of this posting), should the purl be

    pkg:brew/postgresql@12@12.17

or 

    pkg:brew/postgresql@12.17

and let whatever's reading the purl — Homebrew, ostensibly — figure out how that version maps to our major-version formulae?

I speculate only the latter is a valid purl, a safe assumption because I don't think it's a good idea to allow two `@` in the purl.

File: PURL-TYPES.rst [hunk 5]
```diff
@@ -87,6 +87,25 @@ bitnami
...
+- Qualifier ``tap_url``: for taps that are not on GitHub or otherwise require an explicit URL,
+  this is the full URL to the tap.
+- Examples::
+
+      pkg:brew/sqlite@3.43.2
+      pkg:brew/homebrew/core/sqlite@3.43.2
+      pkg:brew/some-org/some-tap/some-app@1.2.3
+      pkg:brew/some-org/some-tap/some-app@1.2.3&tap_url=https://git.example.com/some-org/some-tap.git
```
Comment (colindean) [inline PURL-TYPES.rst:107]: **nitpick (blocking):** 
```suggestion
      pkg:brew/some-org/some-tap/some-app@1.2.3?tap_url=https://git.example.com/some-org/some-tap.git
```

File: PURL-TYPES.rst [hunk 6]
```diff
@@ -87,6 +87,25 @@ bitnami
...
+``brew`` for Homebrew-based packages:
+
+- The default repository (tap) is ``homebrew/core``.
+
+  - The tap syntax is ``https://github.com/{org}/Homebrew-{tap}``, so
+    ``homebrew/core`` corresponds to the URL ``https://github.com/homebrew/homebrew-core``.
+- The ``name`` is the formula name.
+- The ``version`` is the formula version.
+- Qualifier ``tap_url``: for taps that are not on GitHub or otherwise require an explicit URL,
+  this is the full URL to the tap.
```
Comment (colindean) [inline PURL-TYPES.rst:101]: **question (non-blocking):** Should there be another parameter to name the tap? We use the named tap at work, e.g.

    brew tap myco/brewhouse git@git.myco.com:brew/house.git

and the formula, uh, `myco-ctl` could have a purl

    pkg:brew/myco/brewhouse/myco-ctl?tap_url=git@git.myco.com:brew/house.git

and brew would know how to name the tap if it needed to be tapped, or perhaps

    pkg:brew/myco/brewhouse/myco-ctl?tap_url=git@git.myco.com:brew/house.git&tap_name=myco/brewhouse

or more simply

    pkg:brew/myco-ctl?tap_url=git@git.myco.com:brew/house.git&tap_name=myco/brewhouse

**Review by colindean — CHANGES_REQUESTED** [review]
One major nit and a few non-blocking questions and suggestions

File: PURL-TYPES.rst (hunk 2 above)
Comment (woodruffw) [reply PURL-TYPES.rst:97]: Agreed, wordy is good!

File: PURL-TYPES.rst (hunk 1 above)
Comment (matt-phylum) [inline PURL-TYPES.rst:94]: I think care needs to be taken here to distinguish the default tap repository `homebrew/core` from the PURL `repository_url` qualifier, which takes a URL. `homebrew/core` is not a valid URL and should not be put into the `repository_url` qualifier. Besides being confusing to have something that isn't a URL in the URL qualifier, this would break any implementation that expects the repository URL to always be a URL.

There should be an example that shows specifying the repository.

File: PURL-TYPES.rst (hunk 4 above)
Comment (woodruffw) [reply PURL-TYPES.rst:99]: Thanks for calling this out -- my read of the purl spec is that the name and other components are always URL-escaped, e.g. `pkg:brew/postgresql@12@12.17` would actually be encoded as `pkg:brew/postgresql%4012@12.17` in a "wire format" context. So the purl *shoudl* always unambiguously map to a package, without the end client having to do additional disambiguation work.

This is a readability sacrifice, but my (potentially wrong) understanding is that purls are mostly meant to show up only in machine-readable contexts anyways.

File: PURL-TYPES.rst (hunk 6 above)
Comment (woodruffw) [reply PURL-TYPES.rst:101]: My 0.02c would be for your first form:

```
pkg:brew/myco/brewhouse/myco-ctl?tap_url=git@git.myco.com:brew/house.git
```

I _think_ that's unambiguous: the `github.com/{org}/homebrew-{tap}` logic only applies when a `tap_url` isn't present, so I think this is the shortest form that conveys all needed information. But maybe I've overlooked something?

File: PURL-TYPES.rst (hunk 4 above)
Comment (colindean) [reply PURL-TYPES.rst:102]: Sigh, it's not pretty, but if that's the way it needs to be, so be it. If that's the case, I'd recommend adding an example with that…

File: PURL-TYPES.rst [hunk 7]
```diff
@@ -87,6 +87,28 @@ bitnami
...
+    ``https://github.com/homebrew/homebrew-core``.
+  - Any git URL is a valid tap URL (see ``tap_url``)
+
+- The ``name`` is the formula name.
+- The ``version`` is the formula version.
+- Qualifier ``tap_url``: for taps that are not on GitHub or otherwise require an explicit URL,
+  this is the full URL to the tap.
+- Examples::
+
+      pkg:brew/sqlite@3.43.2
```
Comment (colindean) [inline PURL-TYPES.rst:107]: **suggestion:** Based on the encoding comment above:
```suggestion
      pkg:brew/sqlite@3.43.2
      pkg:brew/postgresql%4012@12.17
```

**Review by colindean — APPROVED** [review]

File: PURL-TYPES.rst (hunk 1 above)
Comment (woodruffw) [reply PURL-TYPES.rst]: > There should be an example that shows specifying the repository.

Might be misunderstanding what you mean, but examples 2 through 4 currently show this:

* `pkg:brew/homebrew/core/sqlite@3.43.2` specifies the repository as `homebrew/core`, which is otherwise the implicit default repo;
* `pkg:brew/some-org/some-tap/some-app@1.2.3` specifies the repository as `some-org/some-tap`, which expands to `https://github.com/some-org/homebrew-some-tap`, per Homebrew's tap expansion rules;
* `pkg:brew/some-org/some-tap/some-app@1.2.3?tap_url=https://git.example.com/some-org/some-tap.git` specifies the repository as `https://git.example.com/some-org/some-tap.git`, which then gets tapped with a custom name of `some-org/some-tap` (which could be anything, since the full URL is specified)

File: PURL-TYPES.rst (hunk 4 above)
Comment (woodruffw) [reply PURL-TYPES.rst:102]: Done! Looks like the other examples for other ecosystem purls also contain URL-escaping examples, so we're in-line here 🙂

File: PURL-TYPES.rst (hunk 1 above)
Comment (matt-phylum) [reply PURL-TYPES.rst]: I thought the repository would be in a qualifier like how it is with `repository_url`. If it's in the namespace, that should be called out. Normally, a type spec has a line at the top specifying the default repository URL qualifier value, but `brew` doesn't seem to use the repository URL qualifier. I'm trying to avoid people creating PURLs like `pkg:brew/sqlite@3.43.2?repository_url=homebrew/core` instead.

As it's currently written, there's an explanation of how repositories work, but I don't see an explanation of how the repository is encoded into the PURL. Apparently it goes into the namespace, and there are examples of tap names in the namespace, but the namespace field is never mentioned.

File: PURL-TYPES.rst (hunk 4 above)
Comment (woodruffw) [reply PURL-TYPES.rst:102]: Yep, I added @colindean's example below. Adding an explicit callout also makes sense to me; I'll add some language for that.

File: PURL-TYPES.rst (hunk 1 above)
Comment (woodruffw) [reply PURL-TYPES.rst]: > I thought the repository would be in a qualifier like how it is with repository_url. If it's in the namespace, that should be called out.

Gotcha, I'll do this. I'm not super familiar with the idioms here; is the namespace not the right place to put this? A Homebrew Tap is essentially a package namespace, and the purl spec says this about namespaces:

> **namespace**: some name prefix such as a Maven groupid, a Docker image owner,
  a GitHub user or organization. Optional and type-specific.

Comment (matt-phylum) [reply PURL-TYPES.rst]: I think namespace should be fine as long as it's explained.

Comment (woodruffw) [reply PURL-TYPES.rst]: Ah, I think I see the ambiguity here: Homebrew's taps don't really distinguish between the "repository" and the "namespace," which are distinct concepts in purls. For standard/consistent language purposes, I'll refer to the tap only as a namespace.

**Review by colindean — APPROVED** [review]

**Comment by pombredanne** [comment]
@MikeMcQuaid :wave:  A quick review or ack from you would awesome!

**Comment by pombredanne** [comment]
Thanks!
LGTM overall, though the double `@` in `postgres@16@16.1` name seems a bit weird. Is it possible to run `brew install postgres@16@16.1` or instead `brew install postgres@16.1` or instead `brew install postgres%4016@16.1`

**Comment by matt-phylum** [comment]
According to Stack Overflow, [you cannot install a specific version](https://stackoverflow.com/q/3987683). `pkg:brew/postgresql%4016@16.1` means version 16.1 of Postgres from the postgresql@16 package, which is useful for answering "what's installed?", but apparently not useful for trying to install the same package. The repository apparently doesn't keep old versions: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/p/postgresql@16.rb

**Comment by woodruffw** [comment]
> LGTM overall, though the double `@` in `postgres@16@16.1` name seems a bit weird. Is it possible to run `brew install postgres@16@16.1` or instead `brew install postgres@16.1` or instead `brew install postgres%4016@16.1`

Both `brew install postgres` and `brew install postgres@16` are supported, but the others aren't. Homebrew doesn't _really_ support multiple versions of the same package (per what @matt-phylum said) -- the `@X` syntax is a hack that Homebrew does to allow multiple versions of the same package to exist in a tap when the ecosystem can't all exist on a single version (e.g. OpenSSL, Python, etc.).

TL;DR yes, there is no way to install a _specific_ minor version, e.g. postgres 16.1. `postgres@16` is the full name of a formula, referring specifically to whatever latest version of postgres 16 is packaged by Homebrew. 

Link for reference: https://github.com/Homebrew/homebrew-core/blob/master/Formula/p/postgresql%4016.rb

**Review by MikeMcQuaid — APPROVED** [review]
Looks good to me!

**Comment by woodruffw** [comment]
Gentle ping for review+approval here: this is no longer a blocker in Homebrew's attestation work, but I'd like to get it in so that we can consider it for any future attestation changes, if necessary 🙂

**Review by chenrui333 — APPROVED** [review]

**Review by captn3m0 — APPROVED** [review]
+1 for merging this.

**Comment by matt-phylum** [comment]
I think it's important for this PR to add entries to the test suite. Package names can contain the `@` character and this is not handled correctly by maennchen/purl (https://github.com/maennchen/purl/issues/10) or anchore/packageurl-go (does not accept bug reports). Both fail to handle `pkg:brew/postgres%4016`.

There's also a common misimplementation of the parsing spec which can cause problems here. The spec says "Split the remainder once from the right on '@'. The left side is the remainder. Percent-decode the right side. This is the version." However, some implementations split once from the _left_ on '@'. If somebody writes `pkg:brew/postgres@16@16.1` (non-canonical. the test should ensure the correct canonicalization):
- anchore/packageurl-go, maennchen/purl, sonatype/package-url-java parse it as postgres version 16@16.1 (split from the left instead of the right)
- giterlizzi/perl-URI-PackageURL parse it as version 16 (somehow discarding the 16.1)
- package-url/packageurl-js throws an exception (https://github.com/package-url/packageurl-js/issues/57)

```json
{
  "description": "brew names may contain at signs",
  "purl": "pkg:brew/postgres%4016",
  "canonical_purl": "pkg:brew/postgres%4016",
  "type": "brew",
  "namespace": null,
  "name": "postgres@16",
  "version": null,
  "qualifiers": null,
  "subpath": null,
  "is_invalid": false
},
{
  "description": "brew may contain multiple at signs",
  "purl": "pkg:brew/postgres@16@16.1",
  "canonical_purl": "pkg:brew/postgres%4016@16.1",
  "type": "brew",
  "namespace": null,
  "name": "postgres@16",
  "version": "16.1",
  "qualifiers": null,
  "subpath": null,
  "is_invalid": false
}
```

**Comment by woodruffw** [comment]
Makes sense, although the ship has since sailed on the main backing feature that I needed this for (Homebrew's build provenance feature, which instead uses Homebrew's own wheel filename format for its subject).

As such, I don't have any time allocated for this in the immediate future. I'll try and get back to it when I do have some slack time, but if it's a pressing feature for anyone in the next 1-3 months I have no objection to someone else taking ownership here.

**Comment by colindean** [comment]
💪 I like this test case setup.

File: test-suite-data.json [hunk 8]
```diff
@@ -550,5 +550,113 @@
...
+    "qualifiers": null,
+    "subpath": null,
+    "is_invalid": false
+  },
+  {
+    "description": "brew may specify a namespace that uses the Homebrew default tap lookup mechanism",
+    "purl": "pkg:brew/some-org/some-tap/sqlite@3.43.2?tap_url=https://git.example.com/some-org/some-tap.git",
+    "canonical_purl": "pkg:brew/some-org/some-tap/sqlite@3.43.2?tap_url=https://git.example.com/some-org/some-tap.git",
+    "type": "brew",
+    "namespace": "https://git.example.com/some-org/some-tap.git",
```
Comment (colindean) [inline test-suite-data.json:583]: **suggestion (blocking):** Per [this comment](https://github.com/package-url/purl-spec/pull/281#discussion_r1421046349), I got this case wrong.

```suggestion
    "namespace": "some-org/some-tap",
```

The namespace should always be what's in first part, and it's up to the tooling reading this and calling `brew` to pass the explicit `tap_url` qualifier as the git URL for cloning.

**Review by colindean — CHANGES_REQUESTED** [review]
@woodruffw I got one wrong upon reviewing the unresolved comments. Please merge this suggestion.

@matt-phylum I think the rest of the comments can be resolved.

**Review by colindean — APPROVED** [review]

**Comment by tonylturner** [comment]
Is there an update on this issue? I came here to submit a similar issue and was glad to see this but it's been open for 9 months now. Any ETA on review?

**Comment by woodruffw** [comment]
From the perspective of the Homebrew upstream, this has been done and is ready for final review/merge since June. It'd be nice to have a repository owner to shepherd this, although I'm not exactly sure who that'd be (@pombredanne perhaps? :slightly_smiling_face:)

**Review by jkowalleck — CHANGES_REQUESTED** [review]
please also remove `bre` from the list "Other candidate types to define"

**Comment by woodruffw** [comment]
> please also remove `bre` from the list "Other candidate types to define"

Done with f8fd63e; PTAL.

**Comment by johnmhoran** [comment]
Thank you for your PR @woodruffw.  When you have the chance, could you please resolve the conflicts referred to below?

**Comment by woodruffw** [comment]
This should be good to go again thanks to @colindean -- @johnmhoran @jkowalleck PTAL

**Comment by colindean** [comment]
@johnmhoran @jkowalleck Is there any chance this can be merged in the next few days?

**Comment by jkowalleck** [comment]
need to do some detailed research, before i can actually approve the PR.
Sorry, will not have time for that before mid January.

**Comment by colindean** [comment]
Thanks for that timeline clarity and for the depth of review. The next great window for this would be in time for Homebrew's annual meeting on February 3.

**Comment by colindean** [comment]
Hello, purl came up in a discussion internal to my employer (not Homebrew) and that reminded me to check in on this. How can I or the Homebrew team facilitate your research?

**Review by jkowalleck — CHANGES_REQUESTED** [review]
This whole namespace section is very vague and unclear, and the related implications are just confusing.

I suggest rethinking the entire situation, and draw a much clearer and simpler picture.
My proposal:
1. Remove this custom `tap_url` qualifier idea.  
   Replace it with the well-known qualifier `repository_url`.
   https://github.com/package-url/purl-spec/blob/65d98c4a6f8520b2fe006dff932e1508fc1b8317/PURL-SPECIFICATION.rst?plain=1#L426-L430
1. describe what the `repository_url` is meant to be.  
   Is it a git repo or just a file host?  Which URL schema is allowed here? `git+https:`? `git+ssh:`? And so on ...  
   From the docs:
   > With URL specified, tap a formula repository from anywhere, using any
transport protocol that git(1) handles. The one-argument form of tap
simplifies but also limits. This two-argument command makes no assumptions, so
taps can be cloned from places other than GitHub and using protocols other than
HTTPS, e.g. SSH, git, HTTP, FTP(S), rsync.
1. Get rid of this entire "namespace" idea - it just does not hold.  
   "namespaces" only exist in user-land and are just not worth much. For example, Alice could tap a namespace `foo/bar` with a certain URL, and Bob could tap a namespace `foo/baz` with the same URL, right?
1. Since namespaces are gone, all that remains to distinguish package repo sources is `repository_url`. So there must be a default value for this -- please call it out explicitly. `https://github.com/Homebrew/homebrew-core.git`? 
1. The current definition of `name` is well thought. I would not change it and leave it as is, including the example; better remind users on existing spec, especially the percent-encoding of `@` in the name is already described in the PURL spec   
   https://github.com/package-url/purl-spec/blob/65d98c4a6f8520b2fe006dff932e1508fc1b8317/PURL-SPECIFICATION.rst?plain=1#L241
1. Adjust examples and test suite accordingly.

**Comment by pombredanne** [comment]
@woodruffw @colindean let's try to get this through.

**Comment by SMillerDev** [comment]
> Since namespaces are gone, all that remains to distinguish package repo sources is repository_url. So there must be a default value for this -- please call it out explicitly. https://github.com/Homebrew/homebrew-core.git?

FYI, this can be either `https://github.com/Homebrew/homebrew-core.git` (although the consumption is through an API instead of git) or `https://github.com/Homebrew/homebrew-cask.git` is there a way to define that the default is one of multiple values?

File: PURL-TYPES.rst [hunk 9]
```diff
@@ -89,6 +89,32 @@ bitnami
       pkg:bitnami/wordpress@6.2.0?arch=arm64&distro=debian-12
       pkg:bitnami/wordpress@6.2.0?arch=arm64&distro=photon-4
 
+brew
+----
+``brew`` for Homebrew-based packages:
+
+- There is no default package repository; this should be implied either from
```
Comment (pombredanne) [inline PURL-TYPES.rst:96]: @woodruffw is there really no default? By default everything is coming direct from homebrew at https://formulae.brew.sh/ and from these from what I can see:

- https://github.com/Homebrew/homebrew-core/tree/main/Formula
- https://github.com/Homebrew/homebrew-cask/tree/main/Casks

And all packages are served directly from the (password protected) GH package repos?

This may also help if we can simplify the wording about using  tap_url and use instead the repository_url as suggested by @jkowalleck ?

Do we have examples of other public homebrew repos BTW?

Comment (jkowalleck) [reply PURL-TYPES.rst:96]: > Do we have examples of other public homebrew repos BTW?

https://github.com/CycloneDX/homebrew-cyclonedx

Comment (SMillerDev) [reply PURL-TYPES.rst:96]: For a standard installation in Europe for example there is a default of `Homebrew/core` and `Homebrew/cask` which are backed by an API. Behind the Great Firewall people are using mirrors of the API and packages. But people can also opt to use the GitHub repository as a backing for core and cask.

I believe the API can also work for other repositories, but I'm not 100% sure about that. They might need to have a Git repository.

Comment (jkowalleck) [reply PURL-TYPES.rst:96]: > For a standard installation in Europe for example there is a default of `Homebrew/core` and `Homebrew/cask` [...]

nope. `homebrew/cask` is not enabled by default. it might be auto-enabled when needed.
So, https://github.com/Homebrew/homebrew-core is the default. it even calls out to be the default:
> Default formulae for the missing package manager for macOS (or Linux) 


```shellSession
$ docker run --rm -it homebrew/brew brew tap-info --installed
==> Downloading https://formulae.brew.sh/api/formula.jws.json
homebrew/core: Installed
4 commands, 7717 formulae
/home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core (8,073 files, 1GB)
From: https://github.com/Homebrew/homebrew-core
HEAD: 6584e14d8a36b5de72f546d7ea3479d3c494055e
last commit: 5 days ago
```

Comment (SMillerDev) [reply PURL-TYPES.rst:96]: > nope. homebrew/cask is not enabled by default. it might be auto-enabled when needed.

As a homebrew maintainer I can most definitely assure you it is, maybe you tried Homebrew on Linux where not all defaults line up with the main macOS platform?

Comment (jkowalleck) [reply PURL-TYPES.rst:96]: > > nope. homebrew/cask is not enabled by default. it might be auto-enabled when needed.
> 
> As a homebrew maintainer I can most definitely assure you it is, maybe you tried Homebrew on Linux where not all defaults line up with the main macOS platform?

https://docs.brew.sh/Taps makes the impression that [homebrew/core](https://github.com/Homebrew/homebrew-core) is the default repo. 

> If your tap contains a formula that is also present in [homebrew/core](https://github.com/Homebrew/homebrew-core), that’s fine, but you would need to specify its fully qualified name in the form `<user>/<repo>/<formula>` to install your version.


and again: even the repo's description states that it is the default.
> Default formulae for the missing package manager for macOS (or Linux)


PS: i understand that `homebrew/cask` might be auto-enabled on macOS systems, but it is not the default repo, right?

Comment (woodruffw) [reply PURL-TYPES.rst:96]: Homebrew has two things: formulae and casks. Both have defaults; `homebrew-core` is the default for formulae and `homebrew-cask` is the default for casks. Both are used by default to resolve a user's `brew install foo` request, so they share a namespace (in which formulae historically have taken precedence, but the current policy is - IIRC - to minimize any overlap between the two).

Comment (jkowalleck) [reply PURL-TYPES.rst:96]: casks (which are a special kind of formula?) are only instated from a certain repo/tap since the repo/tap was auto-enabled on macOS and the formula/name did not exist in the actual default repo.

again: `https://github.com/Homebrew/homebrew-core` is default repo, right?
the tap docs say so (see above), the repo description says so (see above), 
and also the man page say so: https://github.com/Homebrew/brew/blob/fe4f472c8d55e336ff20a46f91194e2f0d8b1228/manpages/brew.1#L1597
and the code says so: https://github.com/Homebrew/brew/blob/fe4f472c8d55e336ff20a46f91194e2f0d8b1228/Library/Homebrew/api/formula.rb#L33

Comment (woodruffw) [reply PURL-TYPES.rst:96]: No, casks are not a special kind of formula. They're a completely different kind of package entirely: formulae are for builds of open source projects; casks are for application distributions where Homebrew can't provide a build for licensing or logistical reasons.

Homebrew supports two different things, both of which are defaults. If you do `brew install foo` you will get either a formula or a cask, depending on what `foo` is. This is because both homebrew-core and homebrew-cask are enabled by default. So it's not accurate to say that homebrew-core is *the* default repo; it's *one* of two default repos.

I think maybe the documentation is confusing here: the tap documentation is mostly about formulae, but casks share some of the same machinery. The basic version is that `homebrew-cask` is 100% a default repository, and `brew install foo` will install the foo cask by default on macOS with no user changes or configuration needed.

Comment (jkowalleck) [reply PURL-TYPES.rst:96]: Proposal [as described](https://github.com/package-url/purl-spec/pull/281#pullrequestreview-2624336325):

```rst
- The ``name`` is the formula name. Formula names that contain ``@`` must be percent-encoded, such as ``postgresql%4012`` for ``postgres@12``.
- The ``version`` is the formula version.
- Qualifier ``repository_url``: URL to the tap. Default is ``https://github.com/Homebrew/homebrew-core.git``. 
- Examples::

      pkg:brew/sqlite@3.43.2
      pkg:brew/postgresql%4012@12.17
      pkg:brew/some-app@1.2.3
      pkg:brew/some-app@1.2.3?repository_url=https://git.example.com/some-org/some-tap.git
```

Comment (jkowalleck) [reply PURL-TYPES.rst:96]: in case it is possible to distinguish casks from formulas after installation in any way, i'd propose:


```rst
- the ``namespace`` is the package type - valid values: 
  - "formula" - for formula packages
  - "cask" - for cask packages
- The ``name`` is the package name. Formula names that contain ``@`` must be percent-encoded, such as ``postgresql%4012`` for ``postgres@12``.
- The ``version`` is the package version.
- Qualifier ``repository_url``: URL to the tap. 
  - Default for namespace "formula" is ``https://github.com/Homebrew/homebrew-core.git``. 
  - Default for namespace "cask" is ``https://github.com/Homebrew/homebrew-cask.git``. 
- Examples::
      pkg:brew/formula/sqlite@3.43.2
      pkg:brew/formula/postgresql%4012@12.17
      pkg:brew/cask/some-app@1.2.3
      pkg:brew/cask/some-app@1.2.3?repository_url=https://git.example.com/some-org/some-tap.git
```

Comment (SMillerDev) [reply PURL-TYPES.rst:96]: There is a technical way to distinguish between them, but users don't. They install docker and rarely care if it's built from source or not

**Comment by pombredanne** [comment]
@woodruffw I am sorry for the pain... but after the merge of PR #514, PURL tests and defs are now defined in new JSON schema :angel: :innocent: :grin: :
- See  #514

With the new approach... this PR needs to be updated. Sorry for the churn. Thank you for your patience. :heart:

**Comment by woodruffw** [comment]
Sorry folks, this is no longer a blocker for anything I'm currently working on (and hasn't been for a while, we routed around it for Homebrew's provenance feature), so I'm probably not going to be able to justify reviving a standards attempt here. I hope someone else is able to drive this work forwards with a new PR; I'm going to close this one to avoid confusion about its activeness.

**Comment by chenrui333** [comment]
@pombredanne 👋  I re-raise the pr in https://github.com/package-url/purl-spec/pull/796
